provider/ec2/environ.go

```diff
@@ -269,16 +269,30 @@ func (e *environ) AvailabilityZones(ctx context.ProviderCallContext) (corenetwor
 func (e *environ) InstanceAvailabilityZoneNames(ctx context.ProviderCallContext, ids []instance.Id) ([]string, error) {
 	instances, err := e.Instances(ctx, ids)
 	if err != nil && err != environs.ErrPartialInstances {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
-	zones := make([]string, len(instances))
-	for i, inst := range instances {
+
+	return gatherAvailabilityZones(instances), nil
+}
+
+func gatherAvailabilityZones(instances []instances.Instance) []string {
+	zones := make([]string, 0)
+	for _, inst := range instances {
 		if inst == nil {
 			continue
 		}
-		zones[i] = inst.(*amzInstance).AvailZone
+		switch t := inst.(type) {
+		case *amzInstance:
+			zones = append(zones, t.AvailZone)
+		case *sdkInstance:
```

Comment:
We sometimes ended up getting this instance and not the `amzInstance` causing it to panic.